### PR TITLE
Marketing Data Test - Miscellaneous Updates

### PR DESCRIPTION
**Pull request by kevinofferfit**
Misc updates for the marketing data test function including:
1. Converting `T` from numpy array into a pandas DataFrame
2. Downsampling before cross merge
3. Uncomment treatment effects calculations

### CODE REVIEW

File: scripts/marketing_data_test.py [hunk 1]
```diff
@@ -56,7 +56,8 @@ def scale_marketing_data(X: pd.DataFrame, T: pd.DataFrame, log_scale_skew_thresh
 
     # TODO: Update to use log scaling too. Why not?
     tscale = StandardScaler()
-    T = tscale.fit_transform(T)
+    T_scaled = tscale.fit_transform(T)
+    T = pd.DataFrame(T_scaled, index=T.index, columns=T.columns)
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:60]: `T` became an issue down the road, especially after `pd.concat()` with `X` since it's a numpy array object. So I'm converting it back to `pd.DataFrame` right from the beginning.

File: scripts/marketing_data_test.py [hunk 2]
```diff
@@ -77,30 +78,38 @@ def scale_marketing_data(X: pd.DataFrame, T: pd.DataFrame, log_scale_skew_thresh
...
+    :param T: Treatment
     :param collinearity_thresh: Threshold to pass to SelectNonCollinear
-    :return:
+    :return: Pandas dataframe of X and T after removing collinear features
     """
-    logger.info(f"Starting collinearity thresh {collinearity_thresh} selection on X")
+    logger.info(f"Starting collinearity thresh {collinearity_thresh} selection on X...")
     selector = collinearity.SelectNonCollinear(collinearity_thresh)
+    # Remove columns with constant values, fill empty cells with 0, and convert to numpy object.
+    X = X.loc[:, (X != X.iloc[0]).any()].fillna(0)
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:100]: Removing constant features and fill with 0 before calculating collinearity.

File: scripts/marketing_data_test.py [hunk 3]
```diff
@@ -77,30 +78,38 @@ def scale_marketing_data(X: pd.DataFrame, T: pd.DataFrame, log_scale_skew_thresh
...
     mask = selector.get_support()
     X = X.loc[:, mask]
-    logger.info(f"X now {X.shape}")
+    logger.info(f"After removing collinearlity, X now has a shape of: {X.shape}.")
 
-    logger.info(f"Starting collinearity thresh {collinearity_thresh} selection on T")
+    logger.info(f"Starting collinearity thresh {collinearity_thresh} selection on T...")
     selector = collinearity.SelectNonCollinear(collinearity_thresh)
-    selector.fit(T, y)
+    T = T.loc[:, (T != T.iloc[0]).any()].fillna(0)
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:108]: Do the same with above for `T`.

File: scripts/marketing_data_test.py [hunk 4]
```diff
@@ -284,12 +295,20 @@ def marketing_dml_test(
     # treat_est_df = treat_df.drop('treat_id', axis=1)
     # Effect of the treatments trained on
     x_cols = list(X.columns.values)
-    t_cols = list(treat_df.columns.values)
-    t_cols.remove('treat_id')
+    t_cols = list(T.columns.values)
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:298]: We can't use `treat_df` anymore because we removed collinear columns. Switch to `T` instead.

File: scripts/marketing_data_test.py [hunk 5]
```diff
@@ -284,12 +295,20 @@ def marketing_dml_test(
...
     X['x_id']=X.index.values
-    treat_est_combo = X.merge(treat_df, how='cross')
-    logger.info(f"Calculating effects of {len(treat_df)} treatments on audience size {len(X)} "
-                f" total {len(treat_est_combo)} combinations")
+
+    n_sample = 10000
+    if X.shape[0] > n_sample:
+        treat_est_combo = X.sample(n=n_sample).reset_index().merge(treat_df, how='cross')
+    else:
+        treat_est_combo = X.merge(treat_df, how='cross')
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:305]: Add if-else condition in case we are down-sampling with less than `n_sample` value. There is probably a cleaner way to do it, but this works for now.

File: scripts/marketing_data_test.py [hunk 6]
```diff
@@ -318,14 +337,18 @@ def marketing_dml_test(
...
         logger.info("#############################################################################")
 
-        # logger.info(f"Fitting the standard (confounded) model with params {y_params}")
-        # noncausal_est = XGBClassAUCScore(**y_params)
-        # noncausal_est.fit(pd.concat([X,T], axis=1), y)
-        # logger.info("Standard model fit done")
-        # standard_predictions = pd.DataFrame(noncausal_est.predict_proba(treat_est_combo)[:,1])
-        # standard_predictions = pd.concat([standard_predictions,treat_identifiers],axis=1)
-        # logger.info(f"Non-causal model fit is done")
-        # logger.info(f'{pd.DataFrame(standard_predictions).describe()}')
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:328]: The lines above have been uncommented. I made several changes and will explain them below.

File: scripts/marketing_data_test.py [hunk 7]
```diff
@@ -318,14 +337,18 @@ def marketing_dml_test(
...
+        logger.info(f"Fitting the standard (confounded) model with params {y_params}")
+        noncausal_est = XGBClassAUCScore(**y_params)
+        # TODO: Confirm with Carl that redefining treat_est_combo is fine since we removed
+        # collinear features.
+        treat_est_combo = pd.concat([X,T], axis=1)
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:344]: Assign `pd.concat([X,T], axis=1)` as `treat_est_combo` since this is post collinearity removal. In your commented lines above, you have both pre and post removals.

File: scripts/marketing_data_test.py [hunk 8]
```diff
@@ -375,6 +398,9 @@ def marketing_dml_test(
                     for score in fold_scores:
                         output_results.write(f"{score},")
 
+                if isinstance(est.score_, float):
+                    est.score_ = [est.score_]
```
Comment (kevinofferfit) [inline scripts/marketing_data_test.py:402]: This is the bug where `est.score_` is float for my client sample data. So I am converting to list if it's a float.

**Review by kevinofferfit — COMMENTED** [review]
Commented on several changes from Carl's original branch.

**Review by carl-offerfit — APPROVED** [review]
